aws/resource_aws_copy_snapshot.go

```diff
@@ -0,0 +1,176 @@
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCopySnapshotCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	request := &ec2.CopySnapshotInput{
+		SourceRegion:     aws.String(d.Get("source_region").(string)),
+		SourceSnapshotId: aws.String(d.Get("source_snapshot_id").(string)),
+	}
+	if v, ok := d.GetOk("description"); ok {
+		request.Description = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("encrypted"); ok {
+		request.Encrypted = aws.Bool(v.(bool))
+	}
+	if v, ok := d.GetOk("kms_key_id"); ok {
+		request.KmsKeyId = aws.String(v.(string))
+	}
+
+	res, err := conn.CopySnapshot(request)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*res.SnapshotId)
+
+	err = resourceAwsCopySnapshotWaitForAvailable(d.Id(), conn)
+	if err != nil {
+		return err
+	}
+
+	if err := setTags(conn, d); err != nil {
+		log.Printf("[WARN] error setting tags: %s", err)
+	}
+
+	return resourceAwsCopySnapshotRead(d, meta)
+}
+
+func resourceAwsCopySnapshotRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	req := &ec2.DescribeSnapshotsInput{
+		SnapshotIds: []*string{aws.String(d.Id())},
+	}
+	res, err := conn.DescribeSnapshots(req)
+	if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidSnapshotID.NotFound" {
+		log.Printf("Snapshot %q Not found - removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	snapshot := res.Snapshots[0]
+
+	d.Set("description", snapshot.Description)
+	d.Set("owner_id", snapshot.OwnerId)
+	d.Set("encrypted", snapshot.Encrypted)
+	d.Set("owner_alias", snapshot.OwnerAlias)
+	d.Set("volume_id", snapshot.VolumeId)
+	d.Set("data_encryption_key_id", snapshot.DataEncryptionKeyId)
+	d.Set("kms_key_id", snapshot.KmsKeyId)
+	d.Set("volume_size", snapshot.VolumeSize)
+
+	if err := d.Set("tags", tagsToMap(snapshot.Tags)); err != nil {
+		log.Printf("[WARN] error saving tags to state: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsCopySnapshotDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	return resource.Retry(5*time.Minute, func() *resource.RetryError {
+		request := &ec2.DeleteSnapshotInput{
+			SnapshotId: aws.String(d.Id()),
+		}
+		_, err := conn.DeleteSnapshot(request)
+		if err == nil {
+			return nil
+		}
+
+		ebsErr, ok := err.(awserr.Error)
+		if ebsErr.Code() == "SnapshotInUse" {
+			return resource.RetryableError(fmt.Errorf("EBS SnapshotInUse - trying again while it detaches"))
+		}
+
+		if !ok {
+			return resource.NonRetryableError(err)
+		}
+
+		return resource.NonRetryableError(err)
+	})
+}
+
+func resourceAwsCopySnapshotWaitForAvailable(id string, conn *ec2.EC2) error {
+	log.Printf("Waiting for Snapshot %s to become available...", id)
+
+	req := &ec2.DescribeSnapshotsInput{
+		SnapshotIds: []*string{aws.String(id)},
+	}
+	err := conn.WaitUntilSnapshotCompleted(req)
```

Comment:
This function call can result in an loop if incorrect configurations set. Consider a `source_snapshot_id` which is in `us-west-2` and a config which sets `source_region` to `us-east-1`. In this scenario, the snapshot shows as pending via the AWS API for a few minutes (2m40s was observed) at which point the following is returned (I assume until the max retries is reached).

```
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws: <DescribeSnapshotsResponse xmlns="http://ec2.amazonaws.com/doc/2016-11-15/">
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:     <requestId>d6c4ece6-eded-4a90-a5dd-204191d8bdc5</requestId>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:     <snapshotSet>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:         <item>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <snapshotId>snap-0fb4b3336bc1b1493</snapshotId>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <volumeId>vol-ffffffff</volumeId>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <status>error</status>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <statusMessage>Source snapshot is not found</statusMessage>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <startTime>2018-01-23T00:05:06.000Z</startTime>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <progress>0%</progress>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <ownerId>721355382739</ownerId>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <volumeSize>0</volumeSize>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <description/>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:             <encrypted>false</encrypted>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:         </item>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws:     </snapshotSet>
2018-01-22T16:14:56.506-0800 [DEBUG] plugin.terraform-provider-aws: </DescribeSnapshotsResponse>
```

This will need some more work.